### ISSUE TITLE
Feature request: built-in parallelization

### ISSUE THREAD

**Issue by johnnybubonic** [paramiko/paramiko#1151]
It would be fantastic if Paramiko natively supported parallelized multiple connections in much the similar way [pdsh](https://github.com/chaos/pdsh) does.

Presumably, a connection list could be built in dict form via the following:

```
hosts = {namedhost1: {'host': str(SOMEHOST),
                                     'port': int(SOMEPORT),  # if 'port' key not present, 22
                                     'key': KEY_OBJ_or_PATH_TO_PRIVKEY,  # if not specified and no password, try all privkeys
                                     'password': str(PASSWD)},  # not needed if key auth, etc.
              namedhost2: {'host': str(ANOTHERHOST). ....
```
etc.

and then this object could be passed via, say, paramiko.SSHParallelClient().connect(hosts, ...)

thoughts?

**Comment by ploxiln** [comment]
This is intended to be done in a layer above paramiko, e.g. in tools like [Fabric](https://github.com/fabric/fabric)

**Comment by pkittenis** [comment]
There are existing projects that do this with multiple client options - see [parallel-ssh](https://github.com/ParallelSSH/parallel-ssh).

**Comment by johnnybubonic** [comment]
@pkittenis yeah, i actively use pdsh, i just was looking for a pure-python implementation and it made sense to me that it'd be in paramiko. since the majority of people using it, i'd wager, are sysadmins.

**Comment by pkittenis** [comment]
parallel-ssh is python and one of its ssh client options is paramiko. And not really, there are many other uses for a library than interactive cmd line. You can always drive the library via cmd line if that is what you want.

**Comment by johnnybubonic** [comment]
oh, so it is python. interesting. still, i'd expect this to be in paramiko itself.

as for interactive command line, where on earth did you get that assumption from "sysadmin"? i would think "sysadmin" implied the control of a fleet of servers at once. but i'm not looking for a commandline parallelized shell. as i've stated above twice now, i have and use pdsh for that, but it does not let me have access to each host in a paramiko level of control e.g. with channels, segregated stdout/stdin/stderr, and the like.

regardless, before we get off-track, i think the general assumption one can make is that in the majority of paramiko use cases:
- one is using python to automate tasks
- very likely one is automating tasks across multiple servers (or "targets", or "destinations" if you prefer), hence paramiko
- very likely one would want this to be parallelized or chunked instead of serial, as this can take a long time otherwise depending on said task.

i really don't see what's so unreasonable about asking why this isn't a core part of paramiko instead of needing to implement an additional library (or entire **framework**, as is the case of fabric, which is silly for single tasks and absolutely overkill).

**Comment by pkittenis** [comment]
pdsh was specifically referred to which is an interactive tool..

Anyway, I agree fabric is overkill for most simple use cases. Hence the recommendation for parallel-ssh which seems to fit 100% of your use cases. YMMV.
